### PR TITLE
CORTX-28818: Create admin user in config phase

### PR DESCRIPTION
**Pull request by selvakumaar5496**
Signed-off-by: Selvakumar <selva.k.nambi@seagate.com>

# Problem Statement
- Create admin user in config phase by reading hax endpoint from data pod

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: src/setup/rgw.py [hunk 1]
```diff
@@ -122,14 +124,31 @@ def config(conf: MappedConf):
             Rgw._update_rgw_config_with_endpoints(conf, service_endpoints, instance)
             instance = instance + 1
 
-        Log.info('Config phase completed.')
-        return 0
+        # Read Motr HA endpoint from data pod using hctl fetch-fids and update in config file
```
Comment (cdeshmukh) [inline src/setup/rgw.py:127]: Change -> Read Motr HA (Hax) endpoint 

Also add the reason, on why we are doing this way

File: src/setup/rgw.py [hunk 2]
```diff
@@ -122,14 +124,31 @@ def config(conf: MappedConf):
...
+        Log.info('Reading motr_ha_endpoint from data pod')
+        config_path = Rgw._get_cortx_conf(conf, CONFIG_PATH_KEY)
+        hare_config_dir = os.path.join(config_path, 'hare', 'config', Rgw._machine_id)
+        data_pod_hostname = socket.gethostname().replace('server', 'data')
```
Comment (cdeshmukh) [inline src/setup/rgw.py:131]: put this logic in separate function. We may want to improve this later or reuse even somewhere else

File: src/setup/rgw.py [hunk 3]
```diff
@@ -122,14 +124,31 @@ def config(conf: MappedConf):
...
+
+        fetch_fids_cmd = f'hctl fetch-fids -c {hare_config_dir} --node {data_pod_hostname}'
```
Comment (cdeshmukh) [inline src/setup/rgw.py:133]: if this fails, we should try with -all option, to run through all data pod endpoints & try creating user, until success, or failout after trying all

File: src/setup/rgw.py [hunk 4]
```diff
@@ -122,14 +124,31 @@ def config(conf: MappedConf):
...
-        """Create rgw admin user and start rgw service."""
+        rgw_config_path = Rgw._get_rgw_config_path(conf)
+        Rgw._load_rgw_config(Rgw._rgw_conf_idx, f'ini://{rgw_config_path}')
+        Conf.set(Rgw._rgw_conf_idx, \
+            f'client.radosgw-admin>{RgwEndpoint.MOTR_HA_EP.value}', motr_ha_endpoint)
+        Conf.save(Rgw._rgw_conf_idx)
 
-        Log.info('Create rgw admin user and start rgw service.')
+        Log.info(f'Updated motr_ha_endpoint in config file {rgw_config_path}')
         # admin user should be created only on one node.
```
Comment (shriya-deshmukh) [inline src/setup/rgw.py:152]: Should we move user creation in 'init' phase ?

Comment (selvakumaar5496) [reply src/setup/rgw.py:152]: @cdeshmukh Please confirm

File: src/setup/rgw.py (hunk 2 above)
Comment (selvakumaar5496) [reply src/setup/rgw.py:131]: done

File: src/setup/rgw.py (hunk 3 above)
Comment (selvakumaar5496) [reply src/setup/rgw.py]: Done

File: src/setup/rgw.py [hunk 5]
```diff
@@ -129,89 +131,43 @@ def config(conf: MappedConf):
...
+        Rgw._verify_backend_store_value(conf)
+
+        # Read Motr HA(HAX) endpoint from data pod using hctl fetch-fids and update in config file
+        # Use remote hax endpoint running on data pod which will be available during rgw
+        # config phase since data pod starts before server pod.
+        current_data_node = socket.gethostname().replace('server', 'data')
+        status = Rgw._update_hax_endpoint_and_create_admin(conf, current_data_node)
+        if status != 0:
+            machine_ids = Rgw._get_cortx_conf(conf, 'cluster>storage_set[0]>nodes')
+            data_pod_hostnames = [machine_id for machine_id in machine_ids if \
```
Comment (cdeshmukh) [inline src/setup/rgw.py:144]: remove current_data_node from data_pod_hostnames as its already tried with

File: src/setup/rgw.py (hunk 1 above)
Comment (selvakumaar5496) [reply src/setup/rgw.py]: Done

File: src/setup/rgw.py [hunk 6]
```diff
@@ -129,89 +131,43 @@ def config(conf: MappedConf):
             Rgw._update_rgw_config_with_endpoints(conf, service_endpoints, instance)
             instance = instance + 1
 
+        # Before user creation,Verify backend store value=motr in rgw config file.
+        Rgw._verify_backend_store_value(conf)
+
+        # Read Motr HA(HAX) endpoint from data pod using hctl fetch-fids and update in config file
+        # Use remote hax endpoint running on data pod which will be available during rgw
+        # config phase since data pod starts before server pod.
+        current_data_node = socket.gethostname().replace('server', 'data')
```
Comment (cdeshmukh) [inline src/setup/rgw.py:140]: Add another comment
\# Try data pod of same node first & if it doesnt work, try HAX endpoints from all data pods in cluster

File: src/setup/rgw.py [hunk 7]
```diff
@@ -495,7 +454,97 @@ def _generate_ssl_cert(conf: MappedConf):
...
+                rgw_lock_val = Conf.get(rgw_consul_idx, rgw_lock_key)
+                Log.info(f'rgw_lock value - {rgw_lock_val}')
+                # TODO: Explore consul lock - https://www.consul.io/commands/lock
+                if rgw_lock_val is None:
+                    Log.info(f'Setting confstore value for key :{rgw_lock_key}'
+                             f' and value as :{Rgw._machine_id}')
+                    Rgw._load_rgw_config(rgw_consul_idx, consul_url)
+                    Conf.set(rgw_consul_idx, rgw_lock_key, Rgw._machine_id)
+                    Conf.save(rgw_consul_idx)
+                    Log.info('Updated confstore with latest value')
```
Comment (cdeshmukh) [inline src/setup/rgw.py:516]: print lock value here after updating

**Review by shriya-deshmukh — APPROVED** [review]

File: src/setup/rgw.py [hunk 8]
```diff
@@ -495,7 +454,97 @@ def _generate_ssl_cert(conf: MappedConf):
...
+                    continue
+                elif rgw_lock_val == Rgw._machine_id:
+                    Log.info('Found lock acquired successfully hence processing'
+                             ' with RGW admin user creation.')
+                    rgw_lock = True
+                    break
+                elif rgw_lock_val != Rgw._machine_id:
+                    Log.info(
+                        'Skipping rgw user creation, as rgw lock is already'
+                        f' acquired by {rgw_lock_val}')
```
Comment (cdeshmukh) [inline src/setup/rgw.py:527]: Skipping rgw user creation, as rgw lock is already'
                        f' acquired by other machine with id {rgw_lock_val}'

Comment (selvakumaar5496) [reply src/setup/rgw.py:527]: done

File: src/setup/rgw.py (hunk 7 above)
Comment (selvakumaar5496) [reply src/setup/rgw.py:516]: done

File: src/setup/rgw.py (hunk 5 above)
Comment (selvakumaar5496) [reply src/setup/rgw.py:144]: done

File: src/setup/rgw.py (hunk 6 above)
Comment (selvakumaar5496) [reply src/setup/rgw.py:140]: done

**Review by cdeshmukh — APPROVED** [review]

**Pull request merged**